Excel import

RomanBoyars opened the pull request.

RomanBoyars committed:
Импорт из excel

RomanBoyars committed:
забытая либа

ciklop commented on crm/templatetags/html_helper.py:
```diff
@@ -82,6 +82,20 @@ def vk_small_avatar(vk_user_id):
     return photo
 
 
+def vk_user_id(vk_ref):
```
в html_helper этому делать нечего, в шаблонах этот код не используется, по этому его надо вынести в либо в utils либо в vk_utils. 
Ну и да, по хорошему - код vk_small_avatar тоже надо вынести туда, а в def vk_small_avatar оставить только вызов той функциии. Но вот это, уже в рамках отдельного PR

ciklop commented on crm/views/manager/client.py:
```diff
@@ -250,3 +253,56 @@ class SubscriptionDelete(PermissionRequiredMixin, RevisionMixin, DeleteView):
+                phone_raw = row[cell.column_index_from_string(phone_col) - 1].value
+                phone = re.sub("\D", "", str(phone_raw))
+                birthday = row[cell.column_index_from_string(birthday_col) - 1].value
+                m = re.search("vk.com\/(?P<id>([A-Za-z0-9_])+)", row[cell.column_index_from_string(vk_col) - 1].value)
```
уже есть `utils.VK_PAGE_REGEXP` к которому надо применять search

ciklop commented on crm/views/manager/client.py:
```diff
@@ -250,3 +253,56 @@ class SubscriptionDelete(PermissionRequiredMixin, RevisionMixin, DeleteView):
+                birthday = row[cell.column_index_from_string(birthday_col) - 1].value
+                m = re.search("vk.com\/(?P<id>([A-Za-z0-9_])+)", row[cell.column_index_from_string(vk_col) - 1].value)
+                vk_ref = m.group('id')
+                vk = vk_user_id(vk_ref)
```
А вот определение vk id клиента надо сделать butch операцией, так как клиентов может быть тысяча - то столько за раз нам ВК просто не даст сделать запросов.

Т.е. в начале создаем клиентов, и сохраняем список ID клиента, VK строка из файла. Потом одним запросом получаем список всех пользователей, и идём по ответу и дописываем vk id. ВК возвращает список пользователей в том порядке в котором было передано, по этому сопоставление уже будет кто есть кто.

ciklop commented on crm/views/manager/client.py:
```diff
@@ -250,3 +253,56 @@ class SubscriptionDelete(PermissionRequiredMixin, RevisionMixin, DeleteView):
+                vk_ref = m.group('id')
+                vk = vk_user_id(vk_ref)
+
+                client = Client.objects.create(
```
Это очень тонкий момент - ведь если я второй раз загружу один и тот-же файл, у меня задублируются данные. Нужна проверка на существования такой записи ученика. Ну и такая проверка позволит избежать использования транзакции - если функция посредине работы упала, то повторная загрузка файла "досоздаст" недостающих

RomanBoyars committed:
WIP мпорт CSV
Что исправлено:
Опечатки
Работа с VK API
Статистика по импорту
Что осталось:
Поддержка формата CSV

RomanBoyars committed:
Добавлен столбец с балансом

ciklop committed:
Update html_helper.py

ciklop committed:
Merge branch 'master' into excel_import
